File: android/app/src/main/java/com/booksync/player/PlaybackOffsets.kt
package com.booksync.player

/**
 * The two playback offsets, in one place (issue #42).
 *
 * Before this existed the same gesture moved a different distance depending on
 * where you made it: 15s on the phone player, 10s in Android Auto and on the
 * notification, 15s back / 30s forward on the web, and a 2s jump on the reader's
 * text→audio handoff against a stated server default of 10s.
 *
 * There is no cross-platform config channel for these, so each platform keeps its
 * own copy and they are kept equal by hand:
 *
 *  - Android  — this object
 *  - Web      — `SKIP_SECONDS` / `RESUME_REWIND_SECONDS` in `AudioPlayerContext.jsx`
 *  - Server   — `default_rewind_seconds` in `config.py`
 *
 * See `docs/position-sync-contract.md` § Playback offsets before changing either
 * value; `PlaybackOffsetsTest` asserts them literally so a one-sided edit fails CI.
 *
 * The seek helpers are pure so they can be unit-tested away from a live player —
 * the same reason [ContinuousPlaybackLog] is a seeded object rather than a bare
 * timestamp inside the service.
 */
object PlaybackOffsets {

    /** Skip back and skip forward, on every surface. */
    const val SKIP_MS = 30_000L

    /**
     * How far a resume backs up, so playback doesn't pick up mid-word. Also the
     * size of the reader's text→audio handoff jump, which lands you just before
     * the sentence you were reading rather than on top of it.
     */
    const val RESUME_REWIND_MS = 5_000L

    /** Where a resume should land, given where playback was paused. */
    fun resumePosition(currentMs: Long): Long =
        (currentMs - RESUME_REWIND_MS).coerceAtLeast(0L)

    fun skipBackPosition(currentMs: Long): Long =
        (currentMs - SKIP_MS).coerceAtLeast(0L)

    /**
     * [durationMs] is ignored when it is unknown — ExoPlayer reports
     * `C.TIME_UNSET` (a large *negative*) until the media is prepared, and
     * coercing against that would drag the position backwards instead of forwards.
     */
    fun skipForwardPosition(currentMs: Long, durationMs: Long): Long {
        val target = currentMs + SKIP_MS
        return if (durationMs > 0L) target.coerceAtMost(durationMs) else target
    }
}

File: web/src/lib/playbackOffsets.js
/**
 * Playback offsets — the web copy (issue #42).
 *
 * These two numbers must be identical on every surface: Android's
 * `PlaybackOffsets` and the server's `default_rewind_seconds` carry the same
 * ones. See `docs/position-sync-contract.md` § Playback offsets before
 * changing either.
 *
 * They live in their own module rather than inside `AudioPlayerContext` so the
 * reader → audiobook handoff sites can import the rewind without pulling the
 * whole provider (and its `../api` imports) in with it.
 */

// Both transport buttons, same in each direction.
export const SKIP_SECONDS = 30

// Picking up mid-word after a pause is hard to follow, so a resume backs up a
// few seconds first. Also the size of the text → audio handoff jump.
export const RESUME_REWIND_SECONDS = 5

/**
 * The position the reader's "switch to listening" handoff should start at
 * (issue #212).
 *
 * The contract makes the text → audio handoff a resume: it lands the same
 * RESUME_REWIND_SECONDS before the anchor that an unpause does, which is what
 * Android's `epubToAudioText` already subtracts. `play()` itself stays
 * rewind-free — Home/Continue hands it an explicit position and must not creep
 * backwards — so the rewind is applied at the handoff call sites instead.
 *
 * The *stored* anchor is untouched: a saved position describes where you were,
 * not where you resume.
 */
export function handoffPositionMs(positionMs) {
    return Math.max(0, (positionMs || 0) - RESUME_REWIND_SECONDS * 1000)
}
